### DISCUSSION TITLE
Improve input/output consistency of encoders & decoders

### DISCUSSION THREAD

**Discussion by charlie-dufort** [otosense/recode#8] [Ideas]
- `ChunkedEncoder` takes either a _Collection_ or an _Iterator_  of frames, and returns a _Collection_ of bytes.  
- `ChunkedDecoder` can only take a _Collection_ of bytes, and returns a _Collection_ of frames.
- `IterativeDecoder` can only take a _Collection_ of bytes, and returns a _Iterator_ of frames.

I suggest we add a little more consistency between the type of inputs and outputs:
- `ChunkedEncoder` would take either a _Collection_  or an _Iterator_ of frames, and would return either a _Collection_  or an _Iterator_  of bytes, depending on the input type.  
- `ChunkedDecoder` would take either a _Collection_  or an _Iterator_ of bytes, and would return either a _Collection_  or an _Iterator_  of frames, depending on the input type.  
- `IterativeDecoder` would take either a _Collection_  or an _Iterator_ bytes, and would return either an _Iterator_ of frames, or an _Iterator_ of iterators of frames, depending on the input type.  

Implementation and documentation update should take about 4 hrs

**Comment by thorwhalen** [comment]
As far as consistency, for sure. To make it easier to analyze, @charlie-dufort, could you copy paste the signatures of the objects in question, and perhaps a link to their code? 

I would usually suggest a decorator to do this input and output job, separating that concern. Yet theses codecs will usually know a lot of performance pressure (for example, encoding or decoding live multi-channel streams), so efficiency counts, and probably better in this case we do it in-line.

**Reply by charlie-dufort** [reply]
@thorwhalen Here's the [link](https://github.com/otosense/recode/blob/master/recode/base.py) to the module at the top of which you can find the types and signatures of the objects used in recode.  They are not 100% accurate though.. 

I made a [pull request](https://github.com/otosense/recode/pull/9) in which I made the (in-line) modifications I mentioned in my first message. Please review, and if the changes are ok with you, I'll update the tests and documentation before merging.  

There are still a few things that could be made more consistent. I even considered changing the conceptual design a bit and rewriting the module completely without classes. But I eventually refrained :)

Here are the types and signature I propose in the PR:
- Decoded data
```
Sample = Union[int, float]
Frame = Union[Sample, Collection[Sample]]
Frames = Union[Sequence[Frame], Iterator[Frame]]
```
- Encoded data
```
Chunk = ByteString
Chunks = Union[ByteString, Iterator[ByteString]]
```
- Low-level, internal codec
```
FrameToChunk = Callable[[Frame], Chunk]
ChunkToFrame = Callable[[Chunk], Frame]
IterChunkToFrame = Callable[[Chunk], Iterator[Frame]]
```
- High-level, user-facing codec
```
Encoder = Callable[[Frames], Chunks]
Decoder = Callable[[Chunks], Frames]
```
